### PR TITLE
XWIKI-11922: New App from AWM should have Main.WebHome as parent page

### PR DESCRIPTION
**Pull request by gabrielasmeria**
* Since the default space is configurable, I used $services.model method to find out the homepage.

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-test-pageobjects/src/main/java/org/xwiki/appwithinminutes/test/po/AppWithinMinutesHomePage.java [hunk 1]
```diff
@@ -31,11 +31,6 @@
  */
 public class AppWithinMinutesHomePage extends ViewPage
 {
-    /**
-     * The title of the AppWithinMinutes home page.
-     */
-    public static final String TITLE = "App Within Minutes";
```
Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-test-pageobjects/src/main/java/org/xwiki/appwithinminutes/test/po/AppWithinMinutesHomePage.java]: Why is this removed?

File: xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml [hunk 2]
```diff
@@ -614,7 +614,9 @@ xcontext.put('propertyCustomDisplayer', new PropertyCustomDisplayer(xcontext))
       #set($queryString = 'wizard=true')
       #if(!$xwiki.exists($homePageRef))
         #set($homePageTitle = '$services.localization.render(''platform.appwithinminutes.appHomePageTitle'', [$doc.space])')
-        #set($queryString = "$queryString&amp;editor=inline&amp;template=AppWithinMinutes.LiveTableTemplate&amp;AppWithinMinutes.LiveTableClass_0_class=$escapetool.url($doc.fullName)&amp;title=$escapetool.url($homePageTitle)&amp;parent=AppWithinMinutes.WebHome")
+        #set($wikiHomePageRef = $services.model.resolveDocument('', 'default'))
```
Comment (Enygma2002) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: Question: Do we want the default home page or the current home page?

File: xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-test-pageobjects/src/main/java/org/xwiki/appwithinminutes/test/po/AppWithinMinutesHomePage.java (hunk 1 above)
Comment (Enygma2002) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-test-pageobjects/src/main/java/org/xwiki/appwithinminutes/test/po/AppWithinMinutesHomePage.java]: I also see no point in removing this.

File: xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml (hunk 2 above)
Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: What's the difference? Or do you mean "default" as in "the default default, Main.WebHome"?

Comment (Enygma2002) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: Default = `$services.model.resolveDocument('', 'default')`, i.e. "the default homepage of a wiki" / "the default homepage of the default space of a wiki" / etc., according to the model's configured default entities.

Current = `$services.wiki.currentWikiDescriptor.mainPageReference`, i.e. the currently configured homepage of the current wiki. See http://extensions.xwiki.org/xwiki/bin/view/Extension/Wiki+Application#HConfigureWiki

Using the default might not make sense if the homepage is configured to be something else in the current instance.

At the same time, using the current homepage might not make sense when exporting the app and installing it on a vanilla environment.

File: xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-test-pageobjects/src/main/java/org/xwiki/appwithinminutes/test/po/AppWithinMinutesHomePage.java (hunk 1 above)
Comment (gabrielasmeria) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-test-pageobjects/src/main/java/org/xwiki/appwithinminutes/test/po/AppWithinMinutesHomePage.java]: That variable was used just in [these](https://github.com/gabrielasmeria/xwiki-enterprise/commit/339a00d84f3753b956dcd9caa13abbcf7113aa73) 2 functional tests to go to AWM's homepage, by clicking "App Within Minutes" from the breadcrumb.
And in order to have the current wiki homepage as a parent for the newly created app, I deleted "App Within Minutes" element from the breadcrumb (which was the parent for the newly created app). 
So, instead I used gotoPage() method to access AWM's homepage, as you can see in the link above.

File: xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml (hunk 2 above)
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: @Enygma2002 what @gabrielasmeria did was to follow the behaviour of the [logo link](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/src/main/resources/flamingo/companylogo.vm#L29) from the top menu bar. I'm against storing the configured home page in the parent field because the parent field won't be updated when the home page configuration changes. And saving (possibly) lots of pages when changing the home page configuration is not a good idea. Ideally we should be able to set the value of the parent field to a "token" that represents the configured home page whatever that may be. But since we don't have such a "token" the best option is IMO to store the default home page as parent. An administrator can redirect the default home page to the configured home page. He can also display the title of the configured home page in the title of the default home page using Velocity so that the breadcrumb looks nice.

Comment (Enygma2002) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: @mflorea the workaround sounds decent.

Note that the logo should/will use the currently configured homepage though, instead of hardcoding it.

Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: @Enygma2002 I agree about the logo (because the logo link is dynamic, computed on each page load).

File: xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml [hunk 3]
```diff
@@ -614,7 +614,9 @@ xcontext.put('propertyCustomDisplayer', new PropertyCustomDisplayer(xcontext))
...
+        #set($wikiHomePage = $services.model.serialize($wikiHomePageRef, 'local'))
+        #set($queryString = "$queryString&amp;editor=inline&amp;template=AppWithinMinutes.LiveTableTemplate&amp;AppWithinMinutes.LiveTableClass_0_class=$escapetool.url($doc.fullName)&amp;title=$escapetool.url($homePageTitle)&amp;parent=$wikiHomePage")
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: The (default) ``$wikiHomePage`` is going to be URL-safe most of the time but it's still better to URL-encode it to avoid any problems. You can quickly wrap it with ``$escapetool.url`` but a nicer fix would be to call ``$escapetool.url`` only once 2 lines below when the edit URL is created, passing a map that you have created and filled before. This will also help you better format this line (so that it doesn't exceed the 120 chars limit), making it easier to read on GitHub for instance. Ping me if you have questions.

Comment (gabrielasmeria) [inline xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/ClassEditSheet.xml]: Fixed it. Thanks!
